File: maze/Maze.cpp
#include "maze.h"

s21::Maze::Maze(int row, int col) {
    row_ = row;
    col_ = col;
    RandomValue();
}

s21::Maze::Maze(int row, int col, int bias) {
  row_ = row > 0 ? row <= 50 ? row : 50 : 1;
  col_ = col > 0 ? col <= 50 ? col : 50 : 1;

//  Bias = bias;
//  BiasMax = 100;

//  Init();
    RandomValue();
}

void s21::Maze::RandomValue(){
    rightWalls_.resize(row_, std::vector<int>(col_, 0));
    bottomWalls_.resize(row_, std::vector<int>(col_, 0));

    for (int i = 0; i < row_; ++i) {
        for (int j = 0; j < col_; ++j) {
            rightWalls_[i][j] = std::rand() % 2;  // Установка правой стены
            std::cout << rightWalls_[i][j] << " ";
        }
        std::cout << std::endl;
    }

    for (int i = 0; i < row_; ++i) {
        for (int j = 0; j < col_; ++j) {
            bottomWalls_[i][j] = std::rand() % 2;  // Установка правой стены
            std::cout << bottomWalls_[i][j] << " ";
        }
        std::cout << std::endl;
    }
}

void s21::Maze::FindPath(int startPosX, int startPosY, int endPosX, int endPosY){

}

void s21::Maze::Generate(int getRow(), int getCol()) {
    std::vector<std::vector<bool>> maze(getRow(), std::vector<bool>(getCol()));

    int start_row = std::rand() % getRow();
    int start_col = std::rand() % getCol();

    maze[start_row][start_col] = true;

    for (int r = 0; r < getRow(); ++r) {
        for (int c = 0; c < getCol(); ++c) {
            if (!maze[r][c]) {
                continue;
            }

            // выбор направление движения
            int random_ = std::rand() % 4;
            if (random_ == 0 && (r > 0 && !maze[r - 1][c])) {  // вверх
                maze[r - 1][c] = true;
            } else if (random_ == 1 && (r + 1 < getRow() && !maze[r + 1][c])) {  // вниз
                maze[r + 1][c] = true;
            } else if (random_ == 2 && (c + 1 < getCol() && !maze[r][c + 1])) {  // вправо
                maze[r][c + 1] = true;
            } else if (random_ == 3 && (c > 0 && !maze[r][c - 1])) {  // влево
                maze[r][c - 1] = true;
            } else {
                break;
            }

            //  проверка изолированных областей
            bool isoArea = false;
            for (int i = 0; i < getRow(); ++i) {
                for (int j = 0; j < getCol(); ++j) {
                    if (maze[i][j]) {
                        if (checkWall(maze, i, j)) {
                            isoArea = true;
                            break;
                        }
                    }
                }
                if (isoArea) {
                    break;
                }
            }

            if (isoArea) {
                std::cout << "Есть изолированные области" << std::endl;
            } else {
                printMaze(maze);
            }
        }
    }
}

bool s21::Maze::checkWall(const std::vector<std::vector<bool>>& maze, int row, int col) {
    return maze[row][col - 1] || maze[row][col + 1] || maze[row - 1][col] || maze[row][col + 1];
}

void s21::Maze::printMaze(const std::vector<std::vector<bool>>& maze) {
    for (const auto& row : maze) {
        for (const auto& cell : row) {
            if  (cell) {
                 std::cout << "X";
            } else {
                std::cout << " ";
            }
        }
        std::cout << std::endl;
    }
}

File: A1_Maze_C_sharp-1-develop/src/AvaloniaApplication1/AvaloniaApplication1/Models/Maze.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace s21_Maze {
  public class Maze {
    private int _row;
    private int _col;

    private List<int[]> _rightWalls;
    private List<int[]> _bottomWalls;
    public int Row { get => _row; }
    public int Col { get => _col; }
    public int Bias { get; set; }
    public int BiasMax { get; private set; }
    public List<int[]> RightWalls { get => _rightWalls; }
    public List<int[]> BottomWalls { get => _bottomWalls; }
    public List<int[]> Path { get; private set; }
    public int PathLength { get => Path.Count; }

    public Maze(int row = 4, int col = 4, int bias = 50) {
      _row = row > 0 ? row <= 50 ? row : 50 : 1;
      _col = col > 0 ? col <= 50 ? col : 50 : 1;

      Bias = bias;
      BiasMax = 100;

      Init();
    }
    public void SetSize(int rows, int cols) {
      _row = rows;
      _col = cols;
      Init();
    }
    public void ReadFromFile(string path) {
      _bottomWalls.Clear();
      _rightWalls.Clear();
      Path.Clear();
      var file = File.ReadAllLines(path);

      if (file.Length == 0)
        throw new Exception("Incorrect file format.");

      int[] data = file[0].Split(' ').Select(int.Parse).ToArray();
      _row = data[0];
      _col = data[1];

      if (file.Length != _row * 2 + 2)
        throw new Exception("Incorrect file format.");

      for (int i = 0; i < _row; ++i) {
        _bottomWalls.Add(file[i + _row + 2].Split(' ').Select(int.Parse).ToArray());
        _rightWalls.Add(file[i + 1].Split(' ').Select(int.Parse).ToArray());
      }
    }
    private void Init() {
      Path = new List<int[]>();
      _rightWalls = new List<int[]>();
      _bottomWalls = new List<int[]>();
      for (int i = 0; i < Row; ++i) {
        _rightWalls.Add(new int[_col]);
        _bottomWalls.Add(new int[_col]);
        _rightWalls[i][Col - 1] = 1;
      }
      for (int i = 0; i < Col; ++i) {
        _bottomWalls[Row - 1][i] = 1;
      }
    }
    public void Generate() {
      Init();
      int setUniqueNumber = _col;
      var set = new List<int>(_row);
      for (int i = 1; i <= _col; ++i) set.Add(i);
      for (int row = 0; row < _row; row++) {
        if (row > 0)
          for (int i = 0; i < _col; i++)
            if (_bottomWalls[row - 1][i] == 1)
              set[i] = ++setUniqueNumber;

        int[] rand = new int[_col * 2];
        for (int i = 0; i < rand.Length; ++i) rand[i] = new Random().Next(BiasMax + 2);

        for (int i = 0; i < _col; ++i) {
          if (i < _col - 1 && (rand[i] < Bias || set[i + 1] == set[i]))
            _rightWalls[row][i] = 1;
          else if (i < _col - 1)
            set = set.Select(x => x == set[i + 1] ? set[i] : x).ToList();
        }
        for (int i = 0; i < _col; ++i) {
          int allSetLength = set.Count(x => x == set[i]);
          int setLength = set.GetRange(0, i + 1).Count(x => x == set[i]);
          int bootomCount =
              _bottomWalls[row].Skip(i - setLength + 1).Take(setLength).Count(x => x == 1);
          if (rand[i + _col] > Bias) {
            if ((_rightWalls[row][i] != 1) || (!(bootomCount == setLength - 1) && setLength > 1))
              _bottomWalls[row][i] = 1;
          } else if (_rightWalls[row][i] == 1 && setLength > 1 && bootomCount == 0) {
            _bottomWalls[row][i] = 1;
          }
        }
      }

      for (int i = 0; i < _col; ++i) {
        if ((i < _col - 1 && set[i + 1] != set[i]) && (i != _col - 1))
          _rightWalls[_row - 1][i] = 0;
        if (i < _col - 1)
          set = set.Select(x => x == set[i + 1] ? set[i] : x).ToList();
      }
    }
    public void Save(string path) {
      StringBuilder stringBuilder = new StringBuilder();
      stringBuilder.AppendLine(_row + " " + _col);
      List<string> stringRows = new List<string>();
      for (int i = 0; i < _row; i++) {
        List<string> rowValues = new List<string>();
        for (int j = 0; j < _col; j++) {
          rowValues.Add(_rightWalls[i][j].ToString());
        }
        stringRows.Add(string.Join(" ", rowValues));
      }
      stringBuilder.AppendLine(string.Join("\n", stringRows));
      stringBuilder.AppendLine();
      stringRows.Clear();
      for (int i = 0; i < _row; i++) {
        List<string> rowValues = new List<string>();
        for (int j = 0; j < _col; j++) {
          rowValues.Add(_bottomWalls[i][j].ToString());
        }
        stringRows.Add(string.Join(" ", rowValues));
      }
      stringBuilder.Append(string.Join("\n", stringRows));
      File.WriteAllText(path, stringBuilder.ToString());
    }

    public void FindPath(int startPosX, int startPosY, int endPosX, int endPosY) => Path =
        (List<int[]>)new PathFinder(this, startPosX, startPosY, endPosX, endPosY).FindPath();
  }

}
